tests/scripts/thread-cert/Cert_5_1_04_RouterAddressReallocation.py

```diff
@@ -72,124 +98,172 @@ def test(self):
+        #            - Challenge TLV
+        #            - Mode TLV
+        #            - Scan Mask TLV (MUST have E and R flags set)
+        #            - Version TLV
+        #         The DUT MUST make two separate attempts to reconnect to its
+        #         original partition in this manner
+
+        with pkts.save_index():
+            for i in range(1, 3):
+                pkts.filter_wpan_src64(ROUTER_1).\
+                    filter_LLARMA().\
+                    filter_mle_cmd(MLE_PARENT_REQUEST).\
+                    filter(lambda p: {
+                                      CHALLENGE_TLV,
+                                      MODE_TLV,
+                                      SCAN_MASK_TLV,
+                                      VERSION_TLV
+                                      } <= set(p.mle.tlv.type) and\
+                           p.ipv6.hlim == 255 and\
+                           p.mle.tlv.scan_mask.r == 1 and\
+                           p.mle.tlv.scan_mask.e == 1
+                           ).\
+                    must_next()
+
+            # Step 6: Router_1 MUST attempt to attach to any other partition
+            #         within range by sending a MLE Parent Request.
+            #         The following TLVs MUST be present in the MLE Parent Request:
+            #            - Challenge TLV
+            #            - Mode TLV
+            #            - Scan Mask TLV
+            #            - Version TLV
+
+            pkts.filter_wpan_src64(ROUTER_1).\
+                filter_LLARMA().\
+                filter_mle_cmd(MLE_PARENT_REQUEST).\
+                filter(lambda p: {
+                                  CHALLENGE_TLV,
+                                  MODE_TLV,
+                                  SCAN_MASK_TLV,
+                                  VERSION_TLV
+                                  } <= set(p.mle.tlv.type) and\
+                       p.ipv6.hlim == 255 and\
+                       p.mle.tlv.scan_mask.r == 1 and\
+                       p.mle.tlv.scan_mask.e == 0
+                       ).\
+                must_next()
+
+        # Step 7: Router_1 MUST create a new partition and update the following
+        #         values randomly:
+        #             - Partition ID
+        #             - Initial VN_version & VN_stable_version
+        #             - Initial ID sequence number
+
+        pkts.filter_wpan_src64(ROUTER_1).\
+            filter_LLANMA().\
+            filter_mle_cmd(MLE_ADVERTISEMENT).\
+            filter(lambda p:\
+                   p.mle.tlv.leader_data.partition_id !=
+                   _pkt_pt.mle.tlv.leader_data.partition_id and\
+                   p.mle.tlv.leader_data.data_version !=
+                   _pkt_pt.mle.tlv.leader_data.data_version and\
+                   p.mle.tlv.leader_data.stable_data_version !=
+                   _pkt_pt.mle.tlv.leader_data.stable_data_version
+                   ).\
```

Comment:
I think this has a tiny chance to fail falsely if the new random data versions are same as the old data versions by chance. Is it possible to fix the false fails?

Reply:
It could be. But from the packet verification,  I think checking the data versions with previous is the way to judge the new data versions.

Reply:
Currently, the chance of false fail in 10 runs is 7.5%, which is pretty high?

I would suggest to change as below to reduce the fail chance. 
```suggestion
            filter(lambda p:\
                   p.mle.tlv.leader_data.partition_id !=
                   _pkt_pt.mle.tlv.leader_data.partition_id and\
                   not (p.mle.tlv.leader_data.data_version ==
                   _pkt_pt.mle.tlv.leader_data.data_version and\
                   p.mle.tlv.leader_data.stable_data_version ==
                   _pkt_pt.mle.tlv.leader_data.stable_data_version)
                   ).\
```